lib/ast/AST.cpp

```diff
@@ -75,10 +283,83 @@ void KOREModule::addDeclaration(KOREDeclaration *Declaration) {
 
 void KOREDefinition::addModule(KOREModule *Module) {
   modules.push_back(Module);
+  for (auto decl : Module->getDeclarations()) {
+    if (auto sortDecl = dynamic_cast<KOREObjectCompositeSortDeclaration *>(decl)) {
+      sortDeclarations.insert({sortDecl->getName(), sortDecl});
+    } else if (auto symbolDecl = dynamic_cast<KOREObjectSymbolDeclaration *>(decl)) {
+      symbolDeclarations.insert({symbolDecl->getSymbol()->getName(), symbolDecl});
+    } else if (auto axiom = dynamic_cast<KOREAxiomDeclaration *>(decl)) {
+      axioms.push_back(axiom);
+    }
+  }
 }
 
 void KOREDefinition::addAttribute(KOREPattern *Attribute) {
-  attributes.push_back(Attribute);
+  if (auto constructor = dynamic_cast<KOREObjectCompositePattern *>(Attribute)) {
+    attributes.insert({constructor->getConstructor()->getName(), constructor});
+    return;
+  }
+  assert(false && "Invalid attribute found");
+}
+
+void KOREDefinition::preprocess() {
+  auto symbols = std::unordered_map<std::string, std::vector<KOREObjectSymbol *>>{};
+  for (auto iter = axioms.begin(); iter != axioms.end();) {
+    auto axiom = *iter;
+    if (!axiom->isRequired()) {
+      iter = axioms.erase(iter);
+    } else {
+      axiom->pattern->markSymbols(symbols);
+      iter++;
+    }
+  }
+  for (auto iter = symbols.begin(); iter != symbols.end(); iter++) {
+    auto entry = *iter;
+    for (auto iter = entry.second.begin(); iter != entry.second.end(); iter++) {
+      KOREObjectSymbol *symbol = *iter;
+      auto decl = symbolDeclarations.lookup(symbol->getName());
+      symbol->instantiateSymbol(decl);
+    }
+  }
+  uint32_t nextSymbol = 0;
+  uint16_t nextLayout = 1;
+  auto instantiations = std::unordered_map<KOREObjectSymbol, uint32_t, HashSymbol>{};
+  auto layouts = std::unordered_map<std::string, uint16_t>{};
+  auto variables = std::unordered_map<std::string, std::pair<uint32_t, uint32_t>>{};
+  for (auto iter = symbols.begin(); iter != symbols.end(); iter++) {
+    auto entry = *iter;
+    uint32_t firstTag = nextSymbol;
+    for (auto iter = entry.second.begin(); iter != entry.second.end(); iter++) {
```

Comment:
Definitely prefer ++i when using iterators. The compiler should optimize i++ to ++i, but it's still semantically nicer.

Reply:
agreed!